### ISSUE TITLE
Reduce PDF Size

### ISSUE THREAD

**Issue by timyounes** [qpdf/qpdf#113]
How to reduce PDF size using QPDF ?

**Comment by jberkenbilt** [comment]
This is a fairly vague question, and it is not something qpdf can always do in general. You can use --object-streams=generate to reduce file sizes to some extent.

**Comment by Nokia808** [comment]
Who to do this exactly "--object-streams=generate" ??

It seem that there are many under the hod of your great program !

**Comment by jberkenbilt** [comment]
PDF 1.5 and newer support object streams, which are a way of encoding objects so that the object definitions themselves can be compressed. For most files, use of object streams slightly reduces the file size. qpdf can preserve object streams in the original file, remove them, or generate them. When it generates them, it puts everything into an object stream that can go there, which usually reduces the file size somewhat.

You can run `qpdf --help` or check the manual for a description of all the available command-line options.

I'm glad you're enjoying qpdf. :-)

**Comment by timyounes** [comment]
Thanks for giving me an answer to my question 👍

**Comment by pegleGrot** [comment]
In PDFs without images, is there a way to further improve compression?

**--object-streams=generate** helps, but I'm trying to reach levels similar to what [jPdfTweak](https://sourceforge.net/projects/jpdftweak/) does with ["Optimize PDF size" + "Use better compression"](http://jpdftweak.sourceforge.net/manual/shot10.png).

**Comment by jberkenbilt** [comment]
In the current version of qpdf, there's nothing you can do to get better compression, but maybe I can add something in an upcoming version to allow better compression to be used to a certain degree. I doubt it would help all that much, but zlib allows tuning compression to be slower and better, and I can add the ability to turn that in in qpdf.

I will re-open this bug to track the possibility of tuning zlib in qpdf 9.

The technical issue here is that, the way compression is implemented right now, adding this will break binary compatibility for the qpdf library. However, I am planning a release soon (probably in June) that will do this for other reasons, and I'm trying to cram in several other breaking changes. This will not break source compatibility, but it will require things that use qpdf to be rebuilt. For CLI users, it will just be another flag. (qpdf has so many flags...)

**Comment by pegleGrot** [comment]
Thanks. I'll keep an eye out.

**Comment by jberkenbilt** [comment]
I have added options in qpdf 9 to globally tweak the zlib compression level and to force qpdf to recompress streams compressed with /FlateDecode. It makes very little difference in the overall size at a significant cost in performance, but it's not much code, and I already added it, so I'll leave it in for qpdf 9.

**Comment by trikko** [comment]
I wonder if it is possible to reduce size recompressing pictures.
Two (not exclusive) ways:
- Allow user to set a jpg quality / png compression / ...
- Resize pictures according to DPI. A 4000x4000 jpeg inside a A4 page is probably unnecessarily large. It would be useful to have a -dpi option. For example: -dpi=72 should resize picture size to a lower resolution with unnoticeable quality loss if you're watching from a screen. -dpi=300 should be ok for printing, and so on.

I guess this would save a lot of space for example when people scan documents and save them as pdf at super-ultra-high resolution.

**Comment by jberkenbilt** [comment]
@trikko If you would like something to happen, please open a new issue. Comments on an old issue that was closed a long time ago are likely to be lost. However, since I have seen this, I am happy to answer your question.

You can already set a compression level for zlib. Images can be compressed in PDF files in multiple ways, but "in the wild", most color or grayscale images that are not jpeg are zlib, and that is also true with other streams. So using `--recompress-flate --compression-level=9 --object-streams=generate` will reduce the size somewhat for most files.

qpdf doesn't currently provide a way to set the jpeg quality, but I suspect that would be an easy enhancement. If you'd like to request that, please open a separate ticket, and I will mark it as an enhancement.

Rescaling images is definitely a good thing to do, and I have actually written some code that does that using a combination of qpdf and libmagick++, though it is part of a larger system. I wouldn't add that functionality to the qpdf library, but qpdf can definitely be part of a solution to that problem. Right now, the code with qpdf to do this is a little cumbersome because of how stream data providers work, but I'm planning an enhancement soon, maybe even for qpdf 10.1, that will allow stream data providers to modify the stream's dictionary, which would make it possible for the new code to dynamically determine the image properties. If I implement this enhancement, I will provide an example that uses this for image resampling since that is actually my use case. However, in order to avoid adding a dependency, I'll probably just implement a trivial pixel averaging resampler since that can be implemented in a few lines of code.

In summary:
* zlib compression level can already be specified
* jpeg quality specification is a good enhancement idea; please open a separate ticket if you want, and feel free to reference this one
* resampling can already be done using a combination of qpdf and an external library, and an enhancement I'm already planning will make it slightly easier to code this and will include an example that shows you more or less how to do it.

**Comment by Nokia808** [comment]
I'm opened - just few seconds ago - request for adding jpeg quality specification - see https://github.com/qpdf/qpdf/issues/488

Regarding re-sampling of image, kindly allow the following options:
- 75 dpi
- 150 dpi
- 300 dpi

& it will be nice to have 200 dpi & 400 dpi options.

But regarding re-sampling  enhancement, I have the following note: it should be available for both coloured images & monochrome (white & black images).

**Comment by imz** [comment]
My experience with reducing the size of my scanned PDFs by reducing the quality of some JPEG images is described at <https://github.com/qpdf/qpdf/issues/712>.

I used a Python module from ocrmypdf, which is based on pikepdf.

**Comment by aadzhanow** [comment]
> For reference purposes, this website mentioned qpdf:
> 
> https://www.baeldung.com/linux/pdf-compress
> 
> They showed this commandline variant:
> 
> ```
> qpdf --compress-streams=y --object-streams=generate document.pdf qpdf_compressed.pdf
> ```
> 
> Interestingly this indeed produces smaller .pdf files (for me at the least) than ghostscript does via =screen setting (or whatever the name was for the lowest DPI resolution).
> 
> Why do I mention this? Well, I am in need to compress around 2000 .pdf files, all of which are just a single page. And simple. So I want to find the optimal way to compress them. Right now qpdf leads before ghostscript and other variants.

`for file in *.pdf; do
    qpdf --object-streams=generate --stream-data=compress "$file" "temp1_$file"
    qpdf --linearize "temp1_$file" "temp2_$file"
    qpdf "temp2_$file" "compressed_$file"
    
    rm "temp1_$file" "temp2_$file"
    
    rm "$file"
    
    mv "compressed_$file" "$file"
done`

This worked for me, reduced the size maybe like 30%

**Comment by pegleGrot** [comment]
Maybe better start under [Discussions](https://github.com/qpdf/qpdf/discussions).
More visibility and a better fit. :)

I think Ghostscript before qpdf may end up more compact, but maybe depends on parameters (haven't delved deep into Ghostscript). jPdfTweak (iText?) also helps, as mentioned above.

**Comment by pegleGrot** [comment]
@dave-yotta 
Difficult to say without seeing the exact commandline, but my guess would be that it's images.
Stored very inefficiently in the original, and transformed into low-quality by Ghostscript.

**Comment by dave-yotta** [comment]
It is definitely this - can qpdf deal with this? gs is slow, but maybe thats due to the compression.

We're using gotenberg and under the hood chrome is inflating the images by a lot.
